### ISSUE TITLE
Trying to understand  CroniterBadDateError: failed to find next date

### ISSUE THREAD

**Issue by lowell80** [taichino/croniter#145]
I'm trying to understand the `CroniterBadDateError` exception I'm getting periodically when trying to find the next/previous date.  Here's an example:

```python
Python 3.8.1 (default, Mar  5 2020, 16:56:54)
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>>
>>> from croniter import croniter
>>> from datetime import datetime, timedelta
>>> cron = "0 13 8 1,4,7,10 wed"
>>> croniter(cron, datetime(2020, 1, 1), day_or=False).get_next(datetime)
datetime.datetime(2020, 1, 8, 13, 0)
>>>
>>> croniter(cron, datetime(2020, 9, 24), day_or=False).get_next(datetime)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lalleman/Sandbox/croniter/src/croniter/croniter.py", line 112, in get_next
    return self._get_next(ret_type or self._ret_type, is_prev=False)
  File "/Users/lalleman/Sandbox/croniter/src/croniter/croniter.py", line 211, in _get_next
    result = self._calc(self.cur, expanded,
  File "/Users/lalleman/Sandbox/croniter/src/croniter/croniter.py", line 431, in _calc
    raise CroniterBadDateError("failed to find next date")
croniter.croniter.CroniterBadDateError: failed to find next date
```

Possibly there's some kind of boundary weirdness, or limit as to how far out the algorithm will search for the next match?  Here's an example where the results are skipped across a year boundary.   Dec 31, 2023 shows no next match, but Jan 1, 2024 shows the next match (which is still over 1 year away).
```python
>>> croniter(cron, datetime(2023, 12, 31), day_or=False).get_next(datetime)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lalleman/Sandbox/croniter/src/croniter/croniter.py", line 112, in get_next
    return self._get_next(ret_type or self._ret_type, is_prev=False)
  File "/Users/lalleman/Sandbox/croniter/src/croniter/croniter.py", line 211, in _get_next
    result = self._calc(self.cur, expanded,
  File "/Users/lalleman/Sandbox/croniter/src/croniter/croniter.py", line 431, in _calc
    raise CroniterBadDateError("failed to find next date")
croniter.croniter.CroniterBadDateError: failed to find next date
>>> croniter(cron, datetime(2024, 1, 1), day_or=False).get_next(datetime)
datetime.datetime(2025, 1, 8, 13, 0)
```

The boundary appears to be between 12:58 (fails) and 12:59 (works).

```python
>>> croniter(cron, datetime(2023, 12, 31, 23, 59), day_or=False).get_next(datetime)
datetime.datetime(2025, 1, 8, 13, 0)
>>> croniter(cron, datetime(2023, 12, 31, 23, 58), day_or=False).get_next(datetime)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lalleman/Sandbox/croniter/src/croniter/croniter.py", line 112, in get_next
    return self._get_next(ret_type or self._ret_type, is_prev=False)
  File "/Users/lalleman/Sandbox/croniter/src/croniter/croniter.py", line 211, in _get_next
    result = self._calc(self.cur, expanded,
  File "/Users/lalleman/Sandbox/croniter/src/croniter/croniter.py", line 431, in _calc
    raise CroniterBadDateError("failed to find next date")
croniter.croniter.CroniterBadDateError: failed to find next date
```

I've observed this behavior for both `get_next()` and `get_prev()`.  If there's some limitation to how far the algorithm will look before "giving up" then that's understandable (it certainly wouldn't be ideal for `get_next()` to be stuck in some infinite loop looking for a non-existent date, for example.)

When I first got this error, the `CroniterBadDateError` made me think that something was wrong with the cron expression itself, as this is one of the exceptions thrown by `expand()` (but, upon further review, it's unclear to me why `CroniterBadDateError` it raised there at all, possibly `expand()` should only raise `CroniterBadCronError`?)

---

Sorry for the rambling.   I guess my questions can be summarized as follows:

1.  Is this expected behavior?
1. If so, is there a general safe rule about the how far out get_next/prev will work reliably?   (In other words, can I be certain that this won't happen if the next/previous date is within 1 year of the given starting date?)
1. Is there any documentation or exception cleanup that would help clear this up?  (I'm glad to help contribute in that regard, with some proper direction.)

Thanks!

**Comment by lowell80** [comment]
So I dug a little deeper on this, and it looks like the explanation in terms of code is this:

`croniter.py:417`

```python
        while abs(year - current_year) <= 1:
```

So that explains the 1 year limit I was seeing.  I also confirmed that increasing this limit allows for matching cron expressions that infrequently occur.  (In the example above, it looks like there can be 5 year gaps between matches)

I'm working on a PR to make this a parameter.   I also noticed that this was also discussed in #34, but there wasn't any reference to it in the docs, so I'll get something in the main README as well.

**Comment by kiorky** [comment]
Yep , there are some spagetthi code, help in a PR would be certainly be reviewed and merged with pleasure if relevant !

**Comment by kiorky** [comment]
I think that your best bet for starting the PR is to add those corner cases as failing tests cases and make them pass as they would have to.

**Comment by kiorky** [comment]
released 0.3.35
